ksm.go

```diff
@@ -0,0 +1,497 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type ksmSetting struct {
+	run                bool
+	pagesPerScanFactor int64
+	scanIntervalMS     uint32
+}
+
+func anonPages() (int64, error) {
+	// We're going to parse meminfo
+	f, err := os.Open(memInfo)
+	if err != nil {
+		return -1, err
+	}
+	defer f.Close()
+
+	scan := bufio.NewScanner(f)
+	for scan.Scan() {
+		line := scan.Text()
+
+		// We only care about anonymous pages
+		if !strings.HasPrefix(line, "AnonPages:") {
+			continue
+		}
+
+		// Extract the before last (value) and last (unit) fields
+		fields := strings.Split(line, " ")
+		value := fields[len(fields)-2]
+		totalMemory, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return -1, fmt.Errorf("Invalid integer")
+		}
+
+		// meminfo gives us kB
+		totalMemory *= 1024
+
+		// Fetch the system page size
+		pageSize := (int64)(os.Getpagesize())
+
+		nPages := totalMemory / pageSize
+		return nPages, nil
+	}
+
+	return 0, fmt.Errorf("Could not compute number of pages")
+}
+
+func (s ksmSetting) pagesToScan() (string, error) {
+	if s.pagesPerScanFactor == 0 {
+		return "", errors.New("Invalid KSM setting")
+	}
+
+	nPages, err := anonPages()
```

Comment:
Just a note - I guess what we really want to know is how many pages have been registered as KSM'able via madvise(MADV_MERGEABLE) - but, afaict you cannot get that information from either /sys/kernel/mm/ksm or from /proc/meminfo :-(  So, system wide AnonPages is the nearest good number we can get. We may end up calculating in some anon pages from the host that are not KSM mergeable, but at least we will not under-estimate.